--- yum-3.2.23-orig/yum/rpmsack.py	2009-05-04 11:38:10.000000000 -0400
+++ yum-3.2.23/yum/rpmsack.py	2009-07-10 06:18:36.000000000 -0400
@@ -174,7 +174,7 @@ class RPMDBPackageSack(PackageSackBase):
         # check provides
         tag = self.DEP_TABLE['provides'][0]
         mi = ts.dbMatch()
-        mi.pattern(tag, rpm.RPMMIRE_GLOB, name)
+        mi.pattern(tag, rpm.RPMMIRE_GLOB, name.encode('us-ascii'))
         for hdr in mi:
             pkg = self._makePackageObject(hdr, mi.instance())
             if not result.has_key(pkg.pkgid):
@@ -197,7 +197,7 @@ class RPMDBPackageSack(PackageSackBase):
         ts = self.readOnlyTS()
         result = {}
         
-        mi = ts.dbMatch('basenames', name)
+        mi = ts.dbMatch('basenames', name.encode('us-ascii'))
         for hdr in mi:
             pkg = self._makePackageObject(hdr, mi.instance())
             if not result.has_key(pkg.pkgid):
@@ -443,6 +443,8 @@ class RPMDBPackageSack(PackageSackBase):
 
         ts = self.readOnlyTS()
         if name is not None:
+	    if type(name) == type(u''):
+	    	name = name.encode('utf8')
             mi = ts.dbMatch('name', name)
         elif arch is not None:
             mi = ts.dbMatch('arch', arch)
