diff -ur rpm-4.4.2.3.old/build/files.c rpm-4.4.2.3.new/build/files.c
--- rpm-4.4.2.3.old/build/files.c	2008-08-31 05:00:07.000000000 -0700
+++ rpm-4.4.2.3.new/build/files.c	2008-08-31 05:12:49.000000000 -0700
@@ -23,7 +23,9 @@
 #define	_RPMFI_INTERNAL
 #include "rpmfi.h"
 
-#include <selinux/selinux.h>
+#ifdef WITH_SELINUX
+    #include <selinux/selinux.h>
+#endif
 
 #define	_RPMTE_INTERNAL
 #include "rpmte.h"
@@ -1156,7 +1158,9 @@
 
     sxfn = rpmGetPath("%{?_build_file_context_path}", NULL);
     if (sxfn != NULL && *sxfn != '\0')
+#ifdef WITH_SELINUX
    	matchpathcon_init(sxfn);
+#endif
 
     for (i = 0, flp = fl->fileList; i < fl->fileListRecsUsed; i++, flp++) {
 	const char *s;
@@ -1339,8 +1343,12 @@
 
 	/* Add file security context to package. */
 	mode_t fmode = (uint_16)flp->fl_mode;
+#ifdef WITH_SELINUX
 	int rc = matchpathcon(flp->fileURL, fmode, &scon);
         if ( rc == 0 && scon != NULL) {
+#else
+        if ( scon != NULL) {
+#endif
 	    (void) headerAddOrAppendEntry(h, RPMTAG_FILECONTEXTS, RPM_STRING_ARRAY_TYPE, &scon, 1);
 	    freecon(scon);
         }
diff -ur rpm-4.4.2.3.old/lib/fsm.c rpm-4.4.2.3.new/lib/fsm.c
--- rpm-4.4.2.3.old/lib/fsm.c	2008-08-31 05:00:07.000000000 -0700
+++ rpm-4.4.2.3.new/lib/fsm.c	2008-08-31 05:13:05.000000000 -0700
@@ -636,7 +636,11 @@
     {
 	security_context_t scon = NULL;
 
+#ifdef WITH_SELINUX
 	if ( matchpathcon(fsm->path, st->st_mode, &scon) == 0 && scon != NULL) {
+#else
+	if ( scon != NULL) {
+#endif
 	    /* Get file security context from patterns. */
 	    fsm->fcontext = scon;
 	} else {
@@ -1342,8 +1346,12 @@
 		    /* Get file security context from patterns. */
 		    if (rpmtsSELinuxEnabled(ts) &&
 			! rpmtsFlags(ts) & RPMTRANS_FLAG_NOCONTEXTS) {
+#ifdef WITH_SELINUX
 			if (matchpathcon(fsm->path, st->st_mode, &scon) == 0 &&
 			    scon != NULL) {
+#else
+			if ( scon != NULL) {
+#endif
 				fsm->fcontext = scon;
 				rc = fsmNext(fsm, FSM_LSETFCON);
 			}
diff -ur rpm-4.4.2.3.old/lib/rpmfi.c rpm-4.4.2.3.new/lib/rpmfi.c
--- rpm-4.4.2.3.old/lib/rpmfi.c	2008-08-31 05:00:07.000000000 -0700
+++ rpm-4.4.2.3.new/lib/rpmfi.c	2008-08-31 05:13:12.000000000 -0700
@@ -16,7 +16,9 @@
 #define	_RPMFI_INTERNAL
 #include "rpmfi.h"
 
-#include <selinux/selinux.h>
+#ifdef WITH_SELINUX
+#include <selinux/selinux.h>
+#endif
 
 #define	_RPMTE_INTERNAL	/* relocations */
 #include "rpmte.h"
@@ -1712,7 +1714,9 @@
     }
 
     /* Read security context patterns. */
+#ifdef WITH_SELINUX
     matchpathcon_init(myfn);
+#endif
 
     /* Compute size of argv array blob, concatenating file contexts. */
     nb = ac * sizeof(*fcnb);
@@ -1725,6 +1729,7 @@
 	mode_t fmode = rpmfiFMode(fi);
 	security_context_t scon;
 
+#ifdef WITH_SELINUX
 	if (matchpathcon(fn, fmode, &scon) == 0) {
 	    fcnb[ac] = strlen(scon) + 1;
 /*@-branchstate@*/
@@ -1736,6 +1741,17 @@
             freecon(scon); 
 /*@=branchstate@*/
 	}
+#else
+    fcnb[ac] = strlen(scon) + 1;
+/*@-branchstate@*/
+    if (fcnb[ac] > 0) {
+    fctxt = xrealloc(fctxt, fctxtlen + fcnb[ac]);
+    memcpy(fctxt+fctxtlen, scon, fcnb[ac]);
+    fctxtlen += fcnb[ac];
+    }
+        freecon(scon); 
+/*@=branchstate@*/
+#endif
 	ac++;
     }
 
diff -ur rpm-4.4.2.3.old/lib/rpminstall.c rpm-4.4.2.3.new/lib/rpminstall.c
--- rpm-4.4.2.3.old/lib/rpminstall.c	2008-08-31 05:00:07.000000000 -0700
+++ rpm-4.4.2.3.new/lib/rpminstall.c	2008-08-31 05:13:21.000000000 -0700
@@ -308,6 +308,7 @@
     if (rpmExpandNumeric("%{?_repackage_all_erasures}"))
 	ia->transFlags |= RPMTRANS_FLAG_REPACKAGE;
 
+#ifdef WITH_SELINUX
     /* Initialize security context patterns (if not already done). */
     if (rpmtsSELinuxEnabled(ts) &&
 	!(ia->transFlags & RPMTRANS_FLAG_NOCONTEXTS)) {
@@ -316,6 +317,7 @@
                     matchpathcon_init(fn);
             }
     }
+#endif
     (void) rpmtsSetFlags(ts, ia->transFlags);
 
     probFilter = ia->probFilter;
diff -ur rpm-4.4.2.3.old/python/rpmts-py.c rpm-4.4.2.3.new/python/rpmts-py.c
--- rpm-4.4.2.3.old/python/rpmts-py.c	2008-08-31 05:00:07.000000000 -0700
+++ rpm-4.4.2.3.new/python/rpmts-py.c	2008-08-31 05:13:30.000000000 -0700
@@ -1199,9 +1199,11 @@
     if (rpmtsSELinuxEnabled(s->ts) &&
 	!(s->ts->transFlags & RPMTRANS_FLAG_NOCONTEXTS)) {
 	const char *fn = rpmGetPath("%{?_install_file_context_path}", NULL);
+#ifdef WITH_SELINUX
 	if (fn != NULL && *fn != '\0') {
                 matchpathcon_init(fn);
 	}
+#endif
 	fn = _free(fn);
     } 
 
