diff -ur configure.ac.old configure.ac
--- configure.ac.old	2008-08-31 06:39:04.000000000 -0700
+++ configure.ac	2008-08-31 06:40:01.000000000 -0700
@@ -1189,9 +1189,9 @@
 
 dnl XXX Choose /usr/lib or /usr/lib64 for library installs.
 MARK64=
-case "${target_cpu}" in
-x86_64*|ppc64*|powerpc64*|sparc64*|s390x*)	MARK64=64 ;;
-esac
+dnl case "${target_cpu}" in
+dnl x86_64*|ppc64*|powerpc64*|sparc64*|s390x*)	MARK64=64 ;;
+dnl esac
 AC_SUBST(MARK64)
 
 dnl Determine the canonical arch-vendor-os for the build machine
@@ -1256,7 +1256,7 @@
 AC_SUBST(LOCALEDIR)
 LIBDIR="`echo $libdir | sed 's-/lib$-/%{_lib}-'`"
 AC_SUBST(LIBDIR)
-RPMCONFIGDIR="`echo ${usrprefix}/lib/rpm`"
+RPMCONFIGDIR="`echo ${libdir}/rpm`"
 AC_DEFINE_UNQUOTED(RPMCONFIGDIR, "$RPMCONFIGDIR",
 	[Full path to rpm global configuration directory (usually /usr/lib/rpm)])
 AC_SUBST(RPMCONFIGDIR)
diff -ur scripts/Makefile.am.old scripts/Makefile.am
--- scripts/Makefile.am.old	2008-08-31 06:47:11.000000000 -0700
+++ scripts/Makefile.am	2008-08-31 06:47:28.000000000 -0700
@@ -27,7 +27,7 @@
 
 all:
 
-configdir = ${prefix}/lib/rpm
+configdir = ${libdir}/rpm
 config_SCRIPTS = \
 	brp-compress brp-python-bytecompile brp-java-gcjcompile brp-redhat \
 	brp-strip brp-strip-comment-note \
diff -ur file/magic/Makefile.am.old file/magic/Makefile.am
--- file/magic/Makefile.am.old	2008-08-31 17:57:48.000000000 -0700
+++ file/magic/Makefile.am	2008-08-31 17:58:01.000000000 -0700
@@ -1,5 +1,5 @@
 #pkgdata_DATA = magic magic.mime magic.mgc magic.mime.mgc
-pkglibdir =  @prefix@/lib/rpm
+pkglibdir =  ${libdir}/rpm
 pkglib_DATA = magic magic.mime magic.mgc magic.mime.mgc
 
 EXTRA_DIST = magic2mime Localstuff Header magic.mime $(magic_FRAGMENTS)
diff -ur file/magic/Makefile.in.old file/magic/Makefile.in
--- file/magic/Makefile.in.old	2008-08-31 18:12:03.000000000 -0700
+++ file/magic/Makefile.in	2008-08-31 18:12:15.000000000 -0700
@@ -55,7 +55,7 @@
 DISTFILES = $(DIST_COMMON) $(DIST_SOURCES) $(TEXINFOS) $(EXTRA_DIST)
 
 #pkgdata_DATA = magic magic.mime magic.mgc magic.mime.mgc
-pkglibdir = @prefix@/lib/rpm
+pkglibdir = $(libdir)/rpm
 ACLOCAL = @ACLOCAL@
 AMTAR = @AMTAR@
 AR = @AR@
