commit acfe923bf21a4402a48b6e357a53a398b107f0b3
Author: Ville Skyttä <ville.skytta@iki.fi>
Date:   Sun Apr 12 21:39:50 2009 +0300

    Fix --with-kde with KDE3 (rhbz#466009).

diff --git a/scripts/find-lang.sh b/scripts/find-lang.sh
index 2dd9e0d..bb25b31 100755
--- a/scripts/find-lang.sh
+++ b/scripts/find-lang.sh
@@ -138,7 +138,7 @@ if [ x"$KDE3_HTML" != x -a -d "$TOP_DIR$KDE3_HTML" ]; then
 find $TOP_DIR$KDE3_HTML -type d|sed '
 s:'"$TOP_DIR"'::
 '"$NO_ALL_NAME$KDE"'s:\(.*/HTML/\)\([^/_]\+\)\(.*/'"$NAME"'/\)::
-'"$NO_ALL_NAME$KDE"'s:\(.*/HTML\)\([^/_]\+\)\(.*/'"$NAME"'\)$:%lang(\2) \1\2\3:
+'"$NO_ALL_NAME$KDE"'s:\(.*/HTML/\)\([^/_]\+\)\(.*/'"$NAME"'\)$:%lang(\2) \1\2\3:
 '"$ALL_NAME$KDE"'s:\(.*/HTML/\)\([^/_]\+\)\(.*/[a-zA-Z0-9.\_\-]\+/\)::
 '"$ALL_NAME$KDE"'s:\(.*/HTML/\)\([^/_]\+\)\(.*/[a-zA-Z0-9.\_\-]\+$\):%lang(\2) \1\2\3:
 s:^\([^%].*\)::
