--- configure.ac	2008-08-31 01:43:41.000000000 -0700
+++ configure.ac	2008-08-31 01:44:01.000000000 -0700
@@ -473,11 +473,11 @@
 else
 # Without pkg-config, we'll kludge in some defaults
   addlib="-lnss3 -lplds4 -lplc4 -lnspr4 -lpthread -ldl"
-  addcppflags="-I/usr/include/nss3 -I/usr/include/nspr4"
+  addcppflags="-I/usr/include/nss -I/usr/include/nspr4"
 fi
 save_CPPFLAGS="$CPPFLAGS"
 CPPFLAGS="$save_CPPFLAGS $addcppflags"
-AC_CHECK_HEADER([nss3/nss.h], [
+AC_CHECK_HEADER([nss/nss.h], [
   AC_CHECK_LIB(nss3, NSS_NoDB_Init, [
     AC_DEFINE(HAVE_LIBNSS, 1, [Define to 1 if you have the 'NSS' library (-lnss3).])
     WITH_NSS_INCLUDE="$addcppflags"
