diff -up rpm-4.4.2.3/build/rpmfc.c.prereq rpm-4.4.2.3/build/rpmfc.c
--- rpm-4.4.2.3/build/rpmfc.c.prereq	2008-04-01 09:47:42.000000000 +0300
+++ rpm-4.4.2.3/build/rpmfc.c	2008-04-01 09:54:54.000000000 +0300
@@ -1488,7 +1488,7 @@ static struct DepMsg_s depMsgs[] = {
 	_notpre(RPMSENSE_SCRIPT_POSTUN), 0 },
   { "Requires",		{ "%{?__find_requires}", NULL, NULL, NULL },
 	-1, -1, RPMTAG_REQUIREFLAGS,	/* XXX inherit name/version arrays */
-	RPMSENSE_FIND_REQUIRES|RPMSENSE_TRIGGERIN|RPMSENSE_TRIGGERUN|RPMSENSE_TRIGGERPOSTUN|RPMSENSE_TRIGGERPREIN, 0 },
+	RPMSENSE_PREREQ, RPMSENSE_PREREQ },
   { "Conflicts",	{ "%{?__find_conflicts}", NULL, NULL, NULL },
 	RPMTAG_CONFLICTNAME, RPMTAG_CONFLICTVERSION, RPMTAG_CONFLICTFLAGS,
 	0, -1 },
diff -up rpm-4.4.2.3/lib/rpmlib.h.prereq rpm-4.4.2.3/lib/rpmlib.h
--- rpm-4.4.2.3/lib/rpmlib.h.prereq	2008-04-01 09:13:04.000000000 +0300
+++ rpm-4.4.2.3/lib/rpmlib.h	2008-04-01 09:47:42.000000000 +0300
@@ -505,8 +505,7 @@ typedef	enum rpmsenseFlags_e {
     RPMSENSE_EQUAL	= (1 << 3),
     RPMSENSE_PROVIDES	= (1 << 4), /* only used internally by builds */
     RPMSENSE_CONFLICTS	= (1 << 5), /* only used internally by builds */
-	/* bit 6 used to be RPMSENSE_PREREQ */
-#define	RPMSENSE_PREREQ	RPMSENSE_ANY
+    RPMSENSE_PREREQ     = (1 << 6),     /*!< @todo Legacy. */
     RPMSENSE_OBSOLETES	= (1 << 7), /* only used internally by builds */
     RPMSENSE_INTERP	= (1 << 8),	/*!< Interpreter used by scriptlet. */
     RPMSENSE_SCRIPT_PRE	= ((1 << 9)|RPMSENSE_PREREQ), /*!< %pre dependency. */
